Add Recipe for PODIO

vvolkl opened the pull request:
This adds a recipe for the PODIO datamodel library, which is used in HEP and was previously in HSF/hep-spack.

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
```
Need to update license

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+    version('00-09-02', sha256='8234d1b9636029124235ef81199a1220968dcc7fdaeab81cdc96a47af332d240',  extension='tar.gz', preferred=True)
+    version('00-09', sha256='3cde67556b6b76fd2d004adfaa3b3b6173a110c0c209792bfdb5f9353e21076f')
+    version('00-08', sha256='9d035a7f5ebfae5279a17405003206853271af692f762e2bac8e73825f2af327')
+    version('develop', git='https://github.com/AIDASOFT/podio.git', branch='master')
```
You can move `git=` above below `url=`. Also, you can name version master.

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+            multi=False,
+            description='Use the specified C++ standard when building.')
+
+    depends_on('cmake')
```
```suggestion
    depends_on('cmake', type='build')
```

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+
+    depends_on('cmake')
+    depends_on('root@6.08.06:')
+    depends_on('python@2.7:')
```
```suggestion
    depends_on('python@2.7:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+    depends_on('cmake')
+    depends_on('root@6.08.06:')
+    depends_on('python@2.7:')
+    depends_on('py-pyyaml')
```
```suggestion
    depends_on('py-pyyaml', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+    # As root is installed as an external package we cannot modify its
+    # setup_dependent_environment function to add the xz lib folder to the
+    # LD_LIBRARY_PATH hence we need to do it here.
+    depends_on('xz', when='^root@6.16:')
```
Can you move this up with the other `depends_on` statements?

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+
+    def setup_dependent_environment(self, spack_env, run_env, dspec):
+        spack_env.set('PODIO', self.prefix)
+
```
This line will cause flake8 to fail

vvolkl replied:
whitespace/tab issue, fixed!

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,79 @@
+        spack_env.prepend_path('LD_LIBRARY_PATH', self.spec['root'].prefix.lib)
+        run_env.prepend_path('LD_LIBRARY_PATH', self.spec['root'].prefix.lib)
+        if 'xz' in self.spec:
+            spack_env.prepend_path('LD_LIBRARY_PATH', self.spec['xz'].prefix.lib)
```
I don't think these are necessary are they? RPATH should take care of things at run-time, and they are already added by the compiler wrappers at build-time.

vvolkl commented:
@adamjstewart thanks for your comments! I have to test again if the LD_LIBRARY_PATH is necessary. It was needed to fix errors in the the ROOT dictionary generation. I'll mark it as [WIP] in the meantime.

drbenmorgan commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,61 @@
+    homepage = "https://github.com/AIDASOFT/podio"
+    url      = "https://github.com/AIDASoft/podio/archive/v00-09-02.tar.gz"
+    git      = "https://github.com/AIDADSOFT/podio.git"
+
```
If you're happy to be be notified automatically when changes are made to the podio package, could you add a `maintainers` line here with your GitHub id? You're also welcome to add my id as this is a HEP package!

drbenmorgan commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,61 @@
+            description='Use the specified C++ standard when building.')
+
+    depends_on('cmake', type='build')
+    depends_on('root@6.08.06:')
```
I think the `depends_on` for `root` and `tbb` will need to account for the `cxxstd` variant to ensure ABI matching. To my knowledge there's no automatic way to do this, but the following is the easiest I've seen:

```python
_cxxstd_values=(14, 17)
variant('cxxstd',
        default='17',
        values=('14', '17'),
        multi=False,
        description='Use the specified C++ standard when building.')

for s in _cxxstd_values:
    depends_on('root@6.08.06: cxxstd=' + s, when='cxxstd=' + s)
    depends_on('tbb cxxstd=' + s, when='cxxstd='+s)

drbenmorgan commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,69 @@
+    version('00-09-02', sha256='8234d1b9636029124235ef81199a1220968dcc7fdaeab81cdc96a47af332d240',  extension='tar.gz', preferred=True)
+    version('00-09', sha256='3cde67556b6b76fd2d004adfaa3b3b6173a110c0c209792bfdb5f9353e21076f')
+    version('00-08', sha256='9d035a7f5ebfae5279a17405003206853271af692f762e2bac8e73825f2af327')
+    version('master', branch='master')
```
I think the `master` should appear at the top of the version list (if the effective "newest")

drbenmorgan commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,69 @@
+
+
+    depends_on('cmake', type='build')
+    depends_on('root@6.08.06:')
```
Can be removed, as the `depends_on` in the `_cxxstd_values` loop handles this.

drbenmorgan commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,69 @@
+    depends_on('root@6.08.06:')
+    depends_on('python@2.7:', type=('build', 'run'))
+    depends_on('py-pyyaml', type=('build', 'run'))
+    depends_on('tbb')
```
Can be removed, as the `depends_on` in the `_cxxstd_values` loop handles this.

vvolkl commented:
Hi @drbenmorgan, @adamjstewart thanks again for the helpful comments. I checked and removed some dependencies -  I couldn't reproduce the issue with xz the original author had, so I removed that. However, the changes to `LD_LIBRARY_PATH` were still necessary - not when building the root dependency with spack, but when I take it from our LCG releases via `packages.yaml`. The error I otherwise get is:  

```
  >> 49    /cvmfs/sft.cern.ch/lcg/releases/LCG_97_FCC_1/ROOT/v6.20.02/x86_64-centos7-gcc8-opt/bin/genreflex: error while loadi
           ng shared libraries: libRIO.so: cannot open shared object file: No such file or directory
  >> 50    make[2]: *** [src/podioDict.cxx] Error 127
```

Since this is an important use case, I'd like to keep this for now.

drbenmorgan commented:
Thanks @vvolkl! I'm just waiting for the CI to complete, but I think we're then good to go unless @adamjstewart has further review comments. 

I'm o.k. with the addition of LD_LIBRARY_PATH to cover the external ROOT use case. I'm not aware if there's a way to add additional environment settings for an external in `packages.yaml` other than through a pre-existing module for it (@gartung, @chissg, did you encounter this in your work on the `root` package?).

adamjstewart commented on var/spack/repos/builtin/packages/podio/package.py:
```diff
@@ -0,0 +1,57 @@
+    maintainers = ['vvolkl', 'drbenmorgan']
+
+    version('master', branch='master')
+    version('00-09-02', sha256='8234d1b9636029124235ef81199a1220968dcc7fdaeab81cdc96a47af332d240',  extension='tar.gz', preferred=True)
```
There should be no need for `extension` or `preferred` here. The extension can be detected from the URL, and Spack already defaults to the latest numerical release.

vvolkl replied:
Done.

drbenmorgan commented:
Hi @vvolkl, apologies I got a bit sidetracked from this. I think the latest Travis errors are related to the changes to the setup of the run vs build environment, so probably a straightforward fix.

vvolkl commented:
@drbenmorgan thanks, I didn't notice `dependent_environment` has to be split as well. Should be ok now.
